### ISSUE TITLE
[phasing 8.5/11] Description-trigger optimization (skill-creator)

### ISSUE THREAD

**Issue by unifylabs-dev** [unifylabs-dev/unify-kit#50]
# Phase 8.5 — Description-trigger optimization (skill-creator)

## Goal

When this phase is done, the `plugins/unifylabs-workflow/skills/handoff/SKILL.md` frontmatter `description:` field has been empirically optimized for trigger accuracy via the `skill-creator:skill-creator` plugin's `scripts/run_loop.py` description-optimization loop. A `~/Projects/unify-kit/.claude/phasing/2026-05-24-handoff-skill-build/p85-workspace/` workspace captures: the 20-query trigger eval set (8-10 should-trigger + 8-10 near-miss should-not-trigger), the human-reviewed final eval JSON, the optimization-loop output (5 iterations max, train/test split per skill-creator), the `best_description` selected by held-out test score, and a side-by-side before/after diff. The optimized description lands in `SKILL.md` only after the founder sees the diff and approves. Body / tags / scripts / hooks / references are **not** in scope — only the frontmatter `description:` string is mutated. Verifiable by: optimizer report shows test-score improvement over baseline, founder-approved diff is committed, SKILL.md still parses as valid YAML frontmatter.

## Cross-phase landscape

- P0 — Foundations + open-item resolution
- P1 — Handoff skill core (markdown only)
- P2 — Handoff skill helpers (scripts + slash commands) [TDD]
- P3 — Context-awareness hook + register [TDD]
- P4 — E2E test handoff skill in non-phasing context
- P5 — Phase-executor addendum + checkpoint shape
- P6 — Extend phasing SKILL.md (orchestrator side)
- P7 — `/phase-continue` command + adapt `launch-terminal.sh` [TDD]
- P8 — E2E test phasing checkpoint flow
- **P8.5 — Description-trigger optimization (skill-creator) ← YOU ARE HERE**
- P9 — Promotion + version bump + symlinks
- P10 — Cutover playbook + drive one real cutover

Inserted between P8 and P9 so the optimized description ships inside the 2.1.0 version bump rather than landing as post-release churn.

## Inputs (required reading)

- This spec + P1 handoff (where the original description was authored) + P4 handoff (where the §7.4 ↔ SKILL.md calibration was confirmed) + master plan.
- `~/.claude/plugins/cache/claude-plugins-official/skill-creator/unknown/skills/skill-creator/SKILL.md` — load via `Skill` tool with `skill-creator:skill-creator`. The "Description Optimization" section (Steps 1-4) is the canonical playbook. Do not deviate.
- `~/.claude/plugins/cache/claude-plugins-official/skill-creator/unknown/skills/skill-creator/assets/eval_review.html` — the HTML template for human-reviewing the eval set (Step 2). Used verbatim; do not hand-roll a review UI.
- `~/.claude/plugins/cache/claude-plugins-official/skill-creator/unknown/skills/skill-creator/scripts/run_loop.py` — the optimization loop entry point. Read briefly so you understand the arg contract; do not modify.
- `~/Projects/unify-kit/plugins/unifylabs-workflow/skills/handoff/SKILL.md` lines 1-14 — current frontmatter (name + description + tags). This is the **input** to optimization; the description string is what gets rewritten.
- `~/Projects/unify-kit/specs/2026-05-23-handoff-skill-design.md` §3 (Slash command surface) + §6.1 (Mode detection signals) + §7.4 (LOAD-BEARING discretion table) — the design-spec semantics the optimized description must still convey. The optimizer will rewrite freely; this phase's job is to ensure the rewrite doesn't drop a load-bearing trigger phrase.

## Decisions already baked in

- **Scope is the `description:` field ONLY.** Body, `tags:`, scripts, hooks, references, addenda, slash commands — all locked. If the optimizer recommends body changes, ignore them; this phase is calibrating the triggering metadata, not redesigning the skill.
- **The optimization loop runs to a hard cap of 5 iterations.** `--max-iterations 5` per skill-creator's default. Stop earlier if test-score plateaus (the loop self-terminates on plateau).
- **`best_description` is selected by held-out TEST score, not train.** The optimizer already does this — codified here so future readers know not to overfit on train.
- **The 20-query eval set requires human sign-off before optimization runs.** Skill-creator's Step 2 mandates this via the HTML reviewer; bad queries → bad descriptions. Do NOT skip the review gate even if the auto-generated queries look reasonable.
- **The model passed to `run_loop.py` is the model powering THIS executor session** (per skill-creator's "Step 3: Run the optimization loop"). Read from system prompt; do not hardcode.
- **Founder must see the before/after diff before commit.** Even if test-score improves, the founder owns the wording — surface the diff via `AskUserQuestion` with Approve / Edit-and-apply / Reject options.
- **Reject is a valid outcome.** If the optimizer's `best_description` measurably improves triggering but drops a phrase the founder values (e.g., explicit slash-command listing, mode-detection callout), the phase succeeds with `Status: complete, action: rejected — current description retained`. The empirical baseline data still has value as a record of "we measured and chose the original."

### Sub-phase 8.5A: Generate the 20-query trigger eval set

1. **Load the skill-creator skill** via the `Skill` tool: `skill-creator:skill-creator`. Internalize the "Description Optimization" section's Step 1 guidance on query realism (concrete details, file paths, mixed lengths, lowercase, typos, casual speech allowed).

2. **Draft 20 queries** as a JSON array. Split:
   - **8-10 should-trigger queries.** Cover at least: explicit `/handoff` invocation; implicit context-rescue language ("getting full, save state", "this convo is getting long, hand off to a fresh session"); mode-specific triggers (mid-brainstorm session, mid-phase-executor); resume-side language ("yesterday I left off mid-something, can you pick it up", "what was I working on"); near-context-pressure framing ("getting close to limits, dump what I've got so a new session can pick up"). Mix formal and casual.
   - **8-10 should-not-trigger near-miss queries.** Cover at least: `/handoff-` adjacent commands a user might type but which mean something else ("hand off this task to engineering"); plain context discussion ("how big is my context right now?"); related-but-distinct asks ("summarize what we discussed", "give me a recap"); other skill territories that share keywords ("save state of this file", "write a session log"). The trickiest near-misses (the optimizer learns the most from these) are queries where a naive keyword-match would trigger but semantically shouldn't.

3. **Write the draft eval JSON** to `.claude/phasing/2026-05-24-handoff-skill-build/p85-workspace/trigger-eval-draft.json` with the skill-creator schema:
   ```json
   [
     {"query": "the user prompt", "should_trigger": true},
     {"query": "another prompt", "should_trigger": false}
   ]
   ```

### Sub-phase 8.5B: Human review of the eval set (skill-creator Step 2)

4. **Render the eval_review.html template** per skill-creator's Step 2: read `assets/eval_review.html`, substitute `__EVAL_DATA_PLACEHOLDER__` (the JSON array verbatim, no quotes — it's a JS variable assignment), `__SKILL_NAME_PLACEHOLDER__` (`handoff`), `__SKILL_DESCRIPTION_PLACEHOLDER__` (current frontmatter description, escaped for HTML). Write to `/tmp/eval_review_handoff.html` and `open` it.

5. **Wait for the founder** to edit queries, toggle should-trigger flags, add/remove entries, then click "Export Eval Set". The file downloads to `~/Downloads/eval_set.json` (check for `eval_set (1).json` etc. — take the most recent).

6. **Copy the reviewed eval set** to `.claude/phasing/2026-05-24-handoff-skill-build/p85-workspace/trigger-eval-reviewed.json`. This is the input to the optimizer.

### Sub-phase 8.5C: Run the optimization loop (skill-creator Step 3)

7. **Snapshot the current SKILL.md** for the before/after diff: `cp plugins/unifylabs-workflow/skills/handoff/SKILL.md p85-workspace/SKILL.md.pre-optimization`.

8. **Resolve the model ID** powering this executor session from the system prompt (`claude-opus-4-7[1m]` per P0's empirical findings; strip `[1m]` if `run_loop.py` rejects bracketed forms — test once with a dry-call if unsure).

9. **Invoke `run_loop.py`** in the background:
   ```bash
   cd <skill-creator base dir>
   python -m scripts.run_loop \
     --eval-set <abs path to trigger-eval-reviewed.json> \
     --skill-path <abs path to plugins/unifylabs-workflow/skills/handoff> \
     --model <model-id> \
     --max-iterations 5 \
     --verbose \
     > <p85-workspace>/run_loop.log 2>&1 &
   ```
   Capture the PID. Periodically `tail -50 run_loop.log` to surface iteration progress to the founder (skill-creator says: "periodically tail the output to give the user updates on which iteration it's on and what the scores look like").

10. **Wait for completion.** The loop opens an HTML report in the browser when done and writes JSON output containing `best_description`. Total runtime: ~30-60min (5 iterations × ~10min each × eval set size). Use `Bash` with `run_in_background: true`; do not block the executor on a `wait`.

### Sub-phase 8.5D: Apply or reject

11. **Parse `best_description`** from the loop's JSON output. Compose the proposed SKILL.md diff: only lines 3-13 of the frontmatter change (description spans lines 3-13 currently; the optimizer may shorten or lengthen it).

12. **Render the diff** for the founder with concrete metrics from the optimizer's report:
    - Baseline trigger accuracy (current description on test queries).
    - Best-iteration trigger accuracy (proposed description on test queries).
    - Iteration count actually used.
    - Train-vs-test delta (overfitting check — if train is much higher than test, flag it).

13. **AskUserQuestion** — single question, 3 options:
    - **Approve and apply** (Recommended if test-score improved ≥10pp without overfitting).
    - **Edit and apply** — founder hand-tweaks the proposed description; phase applies the edited version verbatim.
    - **Reject** — keep current description. Phase still succeeds; workspace remains as audit trail.

14. **On Approve or Edit:**
    - Atomically rewrite SKILL.md frontmatter: read full file, replace description block (lines 3-13 or however many the proposed description spans), write to tmp + rename. Preserve `name:`, `tags:`, body — only the description block changes.
    - Re-verify YAML frontmatter parses: `python3 -c "import yaml,sys; yaml.safe_load(open(sys.argv[1])); print('ok')" SKILL.md`.
    - Commit: `chore(handoff): optimize SKILL.md description for trigger accuracy (skill-creator iter <N>)`.
    - Push: no (consistent with run convention — founder pushes when ready).

15. **On Reject:**
    - No file changes. No commit.
    - Workspace stays in `.claude/phasing/2026-05-24-handoff-skill-build/p85-workspace/` as audit trail.

### Sub-phase 8.5E: Workspace hygiene

16. Leave `p85-workspace/` in place — it's the empirical record for P10's cutover playbook and for any future re-optimization run. Do NOT delete the HTML reviewer, the JSON eval sets, the snapshot, or the loop log.

## Verification

- command `test -f .claude/phasing/2026-05-24-handoff-skill-build/p85-workspace/trigger-eval-reviewed.json && jq '. | length' .../trigger-eval-reviewed.json`: PASS if output is between 15 and 30 (allows the founder to add/remove a few during review).
- command `test -f .../p85-workspace/run_loop.log && grep -c "iteration" .../run_loop.log`: PASS if output ≥1 (loop actually ran).
- command `test -f .../p85-workspace/SKILL.md.pre-optimization`: PASS — snapshot exists for diff audit.
- command `python3 -c "import yaml; yaml.safe_load(open('plugins/unifylabs-workflow/skills/handoff/SKILL.md').read().split('---')[1])"`: PASS — frontmatter still parses as valid YAML.
- command `grep -c "^name: handoff$" plugins/unifylabs-workflow/skills/handoff/SKILL.md`: PASS — output 1 (name field unchanged).
- command `git diff HEAD~1 HEAD -- plugins/unifylabs-workflow/skills/handoff/SKILL.md | grep -cE '^[+-][^+-]' || echo 0`: PASS if Approve/Edit (output >0 = description block changed) OR Reject (output = 0, no commit).
- check `if Approve/Edit: optimizer report shows test-score improvement ≥0pp (no regression)` — render value in handoff Verification table.
- check `if Reject: founder's rejection reason captured in handoff Open questions` — explicit, not implicit.
- review `re-read SKILL.md description block after write` — confirm no `name:` or `tags:` corruption, no body bleed, ends with `---` line before body.

## Out of scope

- Body of SKILL.md (workflow, anatomy, sections).
- `tags:` field (the optimizer might suggest tag changes; ignore them this phase — separate concern).
- Reference files under `skills/handoff/references/`.
- Scripts under `skills/handoff/scripts/`.
- Hook file `hooks/context-awareness.sh`.
- Slash command files under `commands/handoff*.md`.
- Eval JSON under `skills/handoff/evals/handoff-evals.json` (P4 deliverable; different shape, different purpose — not what skill-creator optimizes).
- Running the full skill-creator iteration loop (test cases with/without skill, baseline subagents, viewer feedback). That's a multi-session investment; this phase is description-only.

## What the next phase needs to know

P9 (promotion + version bump) inherits whatever final state SKILL.md is in after P8.5 — optimized description (Approve/Edit path) or original description (Reject path). Either way, the README + CHANGELOG entries P9 writes should reference the canonical SKILL.md description, not the design-spec wording, because the canonical may have diverged. P9 should also reference the `p85-workspace/` audit trail in the CHANGELOG entry if the description was optimized (e.g., "[2.1.0] handoff skill description optimized for trigger accuracy via skill-creator (see `.claude/phasing/2026-05-24-handoff-skill-build/p85-workspace/` for empirical record)") — this surfaces the methodology to future maintainers.

P10 (cutover) is unaffected; the cutover playbook describes how to migrate orchestrators, not how to optimize descriptions.

## Insertion mechanics (one-time, orchestrator-side)

Because this phase is being inserted post-master-plan-lock, the orchestrator handles:

1. **Create GitHub tracking issue** `[phasing 8.5/10] Description-trigger optimization (skill-creator)` (or however the numbering convention shakes out — `[phasing 8.5/11]` is also defensible) on `unifylabs-dev/unify-kit`. Label `phasing`.
2. **Insert run.json entry** at array position 9 (between current P8 entry at index 8 and current P9 entry at index 9, which shifts P9 to index 10 and P10 to index 11). Use `n: 11` (next available integer; `n` is informational), `session_name: phase-8-5-description-optimization-skill-creator`, `status: pending`, `issue_number: <new issue>`.
3. **Run.json comment** (top-level new field `insertion_log` or inline in `task_description`): "P8.5 inserted 2026-05-24 after P8 to fold skill-creator description optimization in before P9 promotion freezes the version."
4. **Spawn this phase** via the standard manual-spawn flow once P8 lands and is verified.

These mechanics are orchestrator-owned — the executor session just consumes the run.json + issue body when it starts.
